tests/integration/devfile/cmd_devfile_status_test.go

```diff
@@ -0,0 +1,375 @@
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(context, "devfile.yaml"))
+
+			helper.CmdShouldPass("odo", "push", "-o", "json", "--project", namespace)
+
+			session := helper.CmdRunner("odo", "component", "status", "-o", "json", "--project", namespace, "--follow")
+
+			utils.WaitForOutputToContain("supervisordStatus", session)
+
+			stdoutContents := string(session.Out.Contents())
+
+			entries, err := utils.ParseMachineEventJSONLines(stdoutContents)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Verify supervisord status is present and correct
+			supervisordEntry := utils.GetMostRecentEventOfType(machineoutput.TypeSupervisordStatus, entries, true).(*machineoutput.SupervisordStatus)
+			count := 0
+			for _, status := range supervisordEntry.ProgramStatus {
+				if status.Program == string(common.DefaultDevfileRunCommand) {
+					Expect(status.Status).To(Equal("RUNNING"))
+				} else if status.Program == string(common.DefaultDevfileDebugCommand) {
+					Expect(status.Status).To(Equal("STOPPED"))
+				} else {
+					Fail(fmt.Sprintf("Unexpected program: %v", status.Program))
+				}
+				count++
+			}
+			Expect(count).To(Equal(2))
+
+			// Kill the node processes within the container, returns true when complete
+			Eventually(func() bool {
+				podName := cliRunner.GetRunningPodNameByComponent(cmpName, namespace)
+
+				contents := helper.GetCliRunner().Exec(podName, namespace, "--", "ps", "-ef")
+
+				pids := []string{}
+				for _, str := range strings.Split(contents, "\n") {
+
+					if strings.Contains(str, "node") || strings.Contains(str, "npm") {
+
+						fields := strings.Fields(str)
+						if len(fields) >= 2 {
+							pids = append(pids, fields[1])
+						}
+					}
+				}
+
+				for _, pid := range pids {
+					helper.GetCliRunner().Exec(podName, namespace, "--", "kill", "-9", pid)
+				}
+
+				// We expect (at least) 2 node processes
+				return len(pids) >= 2
+
+			}, 180, 10).Should(Equal(true))
+
+			// Wait for 'odo component status' to report that the programs are no longer RUNNING (EXITED or STOPPED)
+			Eventually(func() bool {
+				stdoutContents := string(session.Out.Contents())
+				entries, err := utils.ParseMachineEventJSONLines(stdoutContents)
+				if err != nil {
+					return false
+				}
+
+				supervisordStatus := utils.GetMostRecentEventOfType(machineoutput.TypeSupervisordStatus, entries, false).(*machineoutput.SupervisordStatus)
+				if supervisordStatus == nil {
+					return false
+				}
+
+				// All programs should be stopped, because we killed the node processes
+				for _, programStatus := range supervisordStatus.ProgramStatus {
+					if programStatus.Status == "RUNNING" {
+						return false
+					}
+				}
+
+				return true
+			}, 180, 10).Should(Equal(true))
+
+			utils.TerminateSession(session)
+
+		})
+
+		It("Verify that odo component status correctly detects component Kubernetes pods", func() {
```

Comment:
Can we add a test case here to check the status changes get properly reflected when we do an `odo push` that takes down the deployment with a new one?

Reply:
The current design/implementation assumes that the consuming tool (IDE, etc) will restart `odo component status` if the deployment (or any other relevant metadata, like URLs) changes, so this test case would be out of scope. Since the consuming tool knows when the deployment has changed (via `odo push` command output, or just by restarting after every push), it knows when to restart `odo component status`. 

We could definitely add more logic to `odo component status` in the future here (though it will still be necessary to restart in some cases, such as URL addition), but I'd prefer to keep the scope of this item as is, and expand on it as needed in subsequent items.